accepted/0000-link-packages-to-source-and-build.md

```diff
@@ -0,0 +1,514 @@
+- Avoids developer-managed keys (as there's no good way to offer trust to the community given the challenges of distributing public keys). This, along with opt-in signing, adds near-zero initial and ongoing overhead for open source maintainers.
+- Maintainers can opt-in to including build provenance information (where the code lives and how it was built) when publishing using `npm publish` for public packages.
+- Incentivizes maintainers to build in the open because of the strong guarantees that this offers.
+- Verification happens transparently on `npm install` without the need to obtain or manage additional tools or keys.
+- Verification should have a negligible performance impact on `npm install`.
+- Verification should be performed without depending on any third-party systems other than the registry.
+- Compatible with third-party npm clients, e.g. `yarn` and `pnpm`.
+- Should allow third-party npm registries, e.g. GitHub Packages, Artifactory and Verdaccio to follow suit and implement similar interfaces.
+- Should be maintained with >99.9% uptime so that developers are not blocked from publishing new packages.
+- Should allow future extensions to support centrally managed signing authorities such as certificates managed by an enterprise and inner source within air-gapped enterprise environments.
+- Buy-in from the broader open source community.
+
+## Non-goals
+- Support for private/inner-source packages to begin with.
+  - Our primary goal is to link packages hosted on the npm registry to the source and build it originated from, for this to be broadly useful, the source code and build need to be publicly available. We're committed to eventually supporting private use cases.
+- Support for self-hosted CI/CD systems.
+- Mitigate every possible supply chain attack against the npm registry. For example:
+  - Registry/mirror/proxy MITM attacks
+  - Denial of service attacks against the registry or client from a compromised mirror/proxy
+    - Rollback/fast-forward attacks (e.g. providing the wrong version for update)
+    - Freeze/endless data attacks (e.g. preventing updates)
+    - Extraneous dependencies attacks (e.g. tricking the client to install undeclared dependencies)
+  - Malicious maintainer (e.g. [event-stream](https://blog.npmjs.org/post/180565383195/details-about-the-event-stream-incident))
+  - Compromised login credentials (where 2FA can help)
+  - Typosquatting attacks (e.g. `lodahs`), [“dependency confusion”](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610)
+  - Accidental vulnerabilities
+
+## Sigstore as signing solution
+The [Sigstore](https://www.sigstore.dev/) project has been selected as the solution to signing npm packages (see detailed explanation below). It is currently the only working solution that supports our key requirements:
+- Links packages hosted on the npm registry to the source and build they originated from (provenance information).
```

Comment:
other than recording "an npm specifier" and "a github repo and SHA" together, how are these linked? published npm packages already include repo and SHA information in the packument, so how would sigstore's linking be different?

Reply:
This information would come as an OIDC claim from the publishing party and thus verifiable... e.g. the claim of a specific repo would come from the GitHub action itself, not meta data published to the registry from a personal machine

Reply:
Just want to be mindful that there's nothing in this RFC which mentions that Actions/uild envs will become immutable which puts their audibility in question (ex. I can build & publish my package with a trusted builder but then delete that action/workflow - killing the link/trace)

Reply:
Does this mean you _have_ to use GitHub as your source repo?

Reply:
No

Reply:
We're proposing ways to make this information harder to falsify. Currently you can say the repo is anything and it's not validated, so an attacker can easily forge this information.

A part of the solution is using information from the OIDC id token that you can get from a supporting CI/CD system, this includes information about the repo and commit. The whole thing is then signed and can be verified against the CI/CD's public key for example.

Reply:
> nothing in this RFC which mentions that source, CI or build envs will be or will become immutable

Yes this is true. Worth noting though that the entry on Rekor and the stored copy on npm will show an audit trail of where it was originally built from. This might still be useful for post-mortem analysis. Having a broken link could also end being a strong signal to the community that a package is no longer maintained and should not be used.

> Does this mean you _have_ to use GitHub as your source repo?

No, we'll support any provider supported by Sigstore's Fulcio service. Today this only includes GitHub Actions but it's vendor neutral and any provider that meets the following requirements will be supported:
- OIDC ID tokens identifying the current workflow/run/build.
- ID token with a custom audience (needs to be set to `sigstore` for Fulcio to accept it).
- Claims about the code repository, commit, build and actor

Reply:
> I'm not sure if `npm` will want to do this but whether links are still valid is something that can be checked during package/artifact verification.

Yeah maybe this could be optional check that can be performed during verification. We probably don't want to do this by default as it could add significant time to the verification if we're checking 1000s+ links for a typical install.

Another idea would be to send the link though a proxy service on npm, and if this service detects a 404 it updates some state on the package that shows a big red warning next to the link next time you view the package on npmjs.com for example. This might cause more headaches than it's worth though, if requests intermittently fail or if the source repo is down and suddenly a bunch of links are broken and need to be fixed when the service is back up.